### PR TITLE
Separation of Concerns

### PR DESCRIPTION
**Pull request by jhalmen**
This series starts the effort of making it easier to keep an overview of the codebase.

Starting with the most recent work of @cklb this started out as a simple exercise to be able to test PlotCharts, ended up being a little bit more involved.
As PlotCharts need access to an Exporter, this was separated out in a first step, while also streamlining its correct usage: it now handles its own worker thread. Its tests are adapted.

PlotCharts now inherit qtgraph's PlotWidget, and handle everything internally: the mouse pointer coordinates now get sent over a Signal, so that the GUI can show them in the status bar.
Plotting performance can now be tested more easily, as pywisp does not need to be running: a simple test script is included.

### CODE REVIEW

**Comment by cklb** [comment]
Nice. Do you want to merge this before I continue with the remaining tasks or shall I work on this branch?

**Comment by jhalmen** [comment]
however you prefer works fine for me

**Comment by cklb** [comment]
Alright, then I'll just push to this branch.

**Comment by cklb** [comment]
I am now out of ideas how to improve the pyqtgraph rendering for big data sets. Maybe we should reduce the plot refresh rate when working in non-windowed mode for time intervals greater than one or two minutes?

**Comment by jhalmen** [comment]
I think the biggest issue is that pyqtgraph does not allow _appending_ to graph data. This becomes plain when seen graphed out:
- **Moving Window:** True
  While the rendering time stays constant, the cost of shoveling all that data into the PlotChart, only to have its internal `XRange` chop most of it off again, grows linearly.
  If, instead, we could just `append` data, the cost of that would be negligible.image

- **Moving Window:** False
  - **Downsampling:** Offimage
  - **Downsampling:** Peakimage
  - **Downsampling:** Subsampleimage
  - **Downsamping:** Meanimage


`off` vs `peak` doesn't really seem to make it better at all, while `mean` and `subsample` both help.

[PlotDataItem Documentation](https://pyqtgraph.readthedocs.io/en/latest/api_reference/graphicsItems/plotdataitem.html) mentions
> When passing x and y data to [PlotDataItem.setData()](https://pyqtgraph.readthedocs.io/en/latest/api_reference/graphicsItems/plotdataitem.html#pyqtgraph.PlotDataItem.setData), use [numpy.ndarray](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray) instead of python’s built-in lists.

Haven't tried that yet, but I'm not sure how that'd help considering we don't actually have the np.arrays lying around, and would have to convert from python lists each time we shove data around.

**Comment by jhalmen** [comment]
I think the best we can do using qtgraph will be to figure out implementing an adaptive subsampling algorithm on pywisp side, so that the cost of shoveling the data around will not rise linearly with time.
Reimplementing the xrange functionality on our side proves how much this helps
image

Now that PlotChart is its own entity, it would be feasible to dust off my experiments using qcustomplot, which has an [`addData` functionality](https://www.qcustomplot.com/documentation/classQCPGraph.html#ae0555c0d3fe0fa7cb8628f88158d420f) which would enable us to only ever copy new data over.
This would reduce plotting cost greatly.

I do believe however, that that should happen in a different PR, when time permits.

**Comment by cklb** [comment]
> I think the best we can do using qtgraph will be to figure out implementing an adaptive subsampling algorithm on pywisp side, so that the cost of shoveling the data around will not rise linearly with time. Reimplementing the xrange functionality on our side proves how much this helps 
Don't bother implementing one, just copy over pyqtgraphs logic. Having that estimate for the ammount of points we desire the subsampling approach should work on the DataBuffers before we call setData as until then everything is passed by reference.

> Now that PlotChart is its own entity, it would be feasible to dust off my experiments using qcustomplot, which has an [`addData` functionality](https://www.qcustomplot.com/documentation/classQCPGraph.html#ae0555c0d3fe0fa7cb8628f88158d420f) which would enable us to only ever copy new data over. This would reduce plotting cost greatly.
> 
> I do believe however, that that should happens in a different PR, when time permits.
I second that. The only thing I would really like to try is reducing the refresh rate for non windowed plots that stretch over multiple minutes.

**Comment by jhalmen** [comment]
> The only thing I would really like to try is reducing the refresh rate for non windowed plots that stretch over multiple minutes.

This feels like a hack, I don't like it

Dug a bit, found the following code in the [PlotDataItem](https://pyqtgraph.readthedocs.io/en/latest/_modules/pyqtgraph/graphicsItems/PlotDataItem.html) for the `autoDownsample`:
```python
if self.opts['autoDownsample']:
    # this option presumes that x-values have uniform spacing
    if xAllFinite:
        finite_x = x
    else:
        # False: (we checked and found non-finites)
        # None : (we haven't performed a check for non-finites yet)
        finite_x = x[np.isfinite(x)]  # ignore infinite and nan values
    if view_range is not None and len(finite_x) > 1:
        dx = float(finite_x[-1]-finite_x[0]) / (len(finite_x)-1)
        if dx != 0.0:
            width = self.getViewBox().width()
            if width != 0.0:  # autoDownsampleFactor _should_ be > 1.0
                ds_float = max(
                    1.0,
                    abs(
                        view_range.width() /
                        dx /
                        (width * self.opts['autoDownsampleFactor'])
                    )
                )
                if math.isfinite(ds_float):
                    ds = int(ds_float)

    # use the last computed value if our new value is not too different.
    # this guards against an infinite cycle where the plot never stabilizes.
    if math.isclose(ds, self._adsLastValue, rel_tol=0.01):
        ds = self._adsLastValue
    self._adsLastValue = ds
    # downsampling is expensive; delay until after clipping.
```
and then later for the actual downsampling:
```python
if ds > 1:
    if self.opts['downsampleMethod'] == 'subsample':
        x = x[::ds]
        y = y[::ds]
    elif self.opts['downsampleMethod'] == 'mean':
        n = len(x) // ds
        # start of x-values try to select a somewhat centered point
        stx = ds // 2
        x = x[stx:stx + n * ds:ds]
        y = y[:n * ds].reshape(n, ds).mean(axis=1)
    elif self.opts['downsampleMethod'] == 'peak':
        n = len(x) // ds
        x1 = np.empty((n, 2))
        # start of x-values; try to select a somewhat centered point
        stx = ds // 2
        x1[:] = x[stx:stx + n * ds:ds, np.newaxis]
        x = x1.reshape(n * 2)
        y1 = np.empty((n, 2))
        y2 = y[:n * ds].reshape((n, ds))
        y1[:, 0] = y2.max(axis=1)
        y1[:, 1] = y2.min(axis=1)
        y = y1.reshape(n * 2)
```

This could be incorporated into the PlotChart, and the downsampled `x`,`y` values cached until `ds` changes, so there's not only less data shoveling, but also _less work_ because we can be smart about the shown data. qtgraph can't do this, because we reset all the data on every update.

**Comment by cklb** [comment]
> This feels like a hack, I don't like it

The point is to get a responsive GUI. Just think about a simple subsampling approach where we only use every 10th data point since that is what some clever logic (see `ds` form above) told us would be best. Using your suggested (and imho good) internal caching design this will eventually lead to a plot that is only updated on every 10th `set_data` (or whatever) call.

As you can see, the same result could have been obtained by just setting a new duty cycle for the plot refresh hook which is 10 times longer. 

Long story short: Give the PlotChart control over its own refresh cycle and we can get the same behavior without any caching whatsoever.

If this nevertheless feels wrong to you I'm absolutely fine with your proposed caching solution.

**Comment by jhalmen** [comment]
Had some time while traveling, finished this up today: using qtgraph's downsampling algorithms and a PlotChart-internal cache for not having to run that on every frame-update, we can finally cap computation time on plotting large datasets:
- `peak`
  image
- `subsample`
  image
- `mean`
  image
- `off`
  image

while `peak` is slower than `subsample` or `mean`, it is still fast enough to render `25fps` which is what we currently cap the update frequency to, anyway. it does look the best tho (i.e. most closely resembles the actual data). it is therefore made the default.

also rebased this on top of the current `master` branch.

**Comment by cklb** [comment]
Very nice work. Are there any more requirements to meet or can we just merge this?

**Comment by jhalmen** [comment]
I still removed the now unused `InterpolationPoints` config option from the docs and examples, so at least it should be consistent now.

I would welcome some more 'real life' testing before we merge this, to make sure everything still behaves as expected, and there aren't any more regressions (like the one fixed in [53b8627](https://github.com/umit-iace/tool-pywisp/pull/145/commits/53b86274f9ce8fbbf7b610bb4eacaceeee0f6d1f) which i just stumbled upon yesterday even tho it's been there for 3 weeks now..)

**Comment by cklb** [comment]
Alright, I'll apply the changes on one of the lab machines and let it run für a while.

**Comment by cklb** [comment]
Alright. Except for the dict issue no real problems showed up on the test run.

The only problems I could find where related to the exporter, to be precise:
- If an invalid file path or file type is given, the gui while nevertheless happily print "Export successful"
- When the gui is closed while an export process is running it does not seem to wait for the export thread to finish.

Regarding the main issue of freezes / hangups for long simulation times no problems occurred. Everything stayed smooth for 3 open plots containing 10 lines in total, running in non-windowed mode, using the whole screen width and stretching over an interval of 2000 seconds.

Also, running exports during that time did not affect the gui and especially the plot updates.

Concluding, the caching combined with the interpolation works really well :D

**Comment by jhalmen** [comment]
> If an invalid file path or file type is given, the gui while nevertheless happily print "Export successful"

Fixed
> When the gui is closed while an export process is running it does not seem to wait for the export thread to finish.

How do you test this?
I would have thought the Exporter's `__del__` handler would make sure that doesn't happen. It's also possible that a global `Close` event cancels all threads instead of waiting for them to finish.

I now added a simple check and ignore shutdown requests if the exporter is still running. Does that sound like a good solution to you?

**Comment by cklb** [comment]
> > If an invalid file path or file type is given, the gui while nevertheless happily print "Export successful"
> 
> Fixed

That was fast :+1: 
 
> > When the gui is closed while an export process is running it does not seem to wait for the export thread to finish.
> 
> How do you test this? 

My totally scientific "procedure" was to mark the current end time, start an export, immediately close the gui and check if a file was created and if it contains all data up to the end time.
Whether it is possible or worth  transforming that into a unit test, I am not so sure.

> I would have thought the Exporter's `__del__` handler would make sure that doesn't happen. It's also possible that a global `Close` event cancels all threads instead of waiting for them to finish.

So did I. 

> I now added a simple check and ignore shutdown requests if the exporter is still running. Does that sound like a good solution to you?

I'll have a look and happily apply my makeshift procedure again.

File: pywisp/widgets/fileselector.py [hunk 1]
```diff
@@ -14,23 +16,38 @@ def __init__(self, formats):
...
+        self.logger = logging.getLogger("FileSelector")
+
+    def _extFromFilter(self, flt):
+        return re.search(r'\*([.a-z]*)', flt).group(1)
 
     def getSaveFileName(self, text="Export as ..."):
         path = self.st.value("path/export_dir")
         ext = self.st.value("path/export_ext")
-        file = os.path.join(path, "export" + ext)
+        file = os.path.join(path, "export")
```
Comment (cklb) [inline pywisp/widgets/fileselector.py:28]: I would really like to keep the last exported file type as a suggestion for the extension. 
Ideally, the last file type format would also be used as first entry of 'filterStr'.

File: pywisp/widgets/fileselector.py [hunk 2]
```diff
@@ -14,23 +16,38 @@ def __init__(self, formats):
...
+                                                   filterStr, self.formats[ix])
+        if not selpath:
+            self.logger.warn("File selection aborted!")
+            return
+        name, selext = os.path.splitext(selpath)
+        if not selext:
+            selext = self._extFromFilter(selfilt)
+            selpath += selext
+        elif selext not in selfilt:
+            self.logger.error(f"Extension '{selext}' doesn't match used filter. Aborting!")
```
Comment (cklb) [inline pywisp/widgets/fileselector.py:42]: I know that this will rarely happen anyway but if an extension is given, I think we really should not care which filter was selected in the file dialog while doing so. As long as the extension is supported by the exporter, of course.

File: pywisp/gui.py [hunk 3]
```diff
@@ -1244,6 +1244,10 @@ def closeEvent(self, QCloseEvent):
         """
         if self.isConnected:
             self.disconnect()
+        if self.exporter != None:
```
Comment (cklb) [inline pywisp/gui.py:1247]: Instead of ignoring the close event, it would be better to wait for the exporters worker here (thus moving the logic from Exporter.__del__ here) since the timing of the __del__ call seems to be non deterministic. But the handling of the close event is.

I will add a test case that showcases the issue.

**Comment by cklb** [comment]
I added some remarks. Please tell me what you think about the issues. I'll gladly draft some commits to address them if you like.

File: pywisp/gui.py [hunk 4]
```diff
@@ -664,36 +657,13 @@ def _initSettings(self):
         self._addSetting("log_colors", "DEBUG", "#4682B4")
         self._addSetting("log_colors", "NOTSET", "#000000")
 
-        # plot management
```
[the remark this answers is not in the document]
Comment (jhalmen) [reply pywisp/gui.py:667]: I split out the settings handling, I believe this is now cleaner and it works

File: pywisp/gui.py (hunk 3 above)
Comment (jhalmen) [reply pywisp/gui.py:1207]: I think this is a simple (and always correct) way of dealing with this. The user is notified, so knows what to expect, and can just try closing the window later.

Before implementing this, I had tried actually waiting and then continuing to shut down, but none of the attempts actually worked:
- waiting on the thread would just block without ever actually shutting down
- using a timer that would just call the close event again at a later time would raise errors about using things in python that were deleted on the c++ side

but if you work out  a way that works, I'm happy to reconsider

File: pywisp/widgets/fileselector.py (hunk 1 above)
Comment (jhalmen) [reply pywisp/widgets/fileselector.py:22]: The last used format is already the selected one when the FileSelector opens up.

Comment (cklb) [reply pywisp/widgets/fileselector.py:22]: Sorry, I did not realize that. However, I still believe that being verbose about whats happening and putting the last used extension (which matches the last used filter as I now know) directly in the file name surely won't hurt.

**Comment by cklb** [comment]
> I'll comment on the `FileSelector` comments here, as I see them as basically the same issue: different expectations about how it is to be used, and the functionality that we want to enable.
Amen to that.
> 
>     * filename suggestion:
>       In my opinion, this is mostly a gimmick, and mostly useful only for quick tests of the Exporter. As it stands now, if it were to be used, exports would always be overwritten. To export something useful, you currently have to (1) type in a name, (2) click on a different export to have the filename, then click into the name and change it.
>       For the suggestion to be actually useful, it would have to include the date, time, used experiment.
Just to set things clear: I do really not care what the suggested file name part (currently "export") looks like. As you stated correctly, providing something really useful here would mean a lot of work and provide only a little benefit.
The only thing I was trying to talk about is the extension. With the auto-extension feature below, we are adding one anyway, so why not tell the user up front what will be added in the first place?

> 
>     * auto-extension:
>       Given the fact that the filename has to be typed out mostly by hand, I feel it's nice for the user to add the extension for them. Also this is something everybody coming from a windows-background (most students?) will expect.
Let me state it again: I am positive about adding an extension, we should just do it directly in the dialog were we can see what is happening, not after clicking on save.

As I have no access to any windows machine at the moment, I had a look at the file dialogs of firefox and gimp and asked a friendly colleague in the office to try it in excel and word on windows.
The behavior across all programs was the same: The dialog will provide you with a template name and a pre-selected extension (and a matching filter). Changing the filter automatically changes the extension and clicking save does not add something to the full file name but rather saves the thing under the string that was given by the user.
> 
>     * mismatch extension/filter:
>       Now having thought about it for a bit more, I can actually agree on this that if the user already took the time to type out the extension we can accept it if we know about it. Have the patch lying around, will push it in a second
+1

File: pywisp/gui.py (hunk 3 above)
Comment (cklb) [reply pywisp/gui.py:1207]: You are right, this way definitively works. Furthermore I did not know that you already invested so much time into the issue.

As shown by the test case, overloading the `__del__` call does in fact not stop the deletion, so I moved the logic there into the `closeEvent` handler on friday and on the two test systems that I have available it worked as expected. That was actually the reason for my initial remark on this.
As for why this approach did not work on your machine I have no idea and before taking the chance of the same thing happening elsewhere I think staying with the solution you provided is the best way to go forward.

I think we can safely remove the  `__del__`  handler, though.

**Comment by jhalmen** [comment]
> Changing the filter automatically changes the extension

~This is the part we can't implement currently. This is why i'm reluctant about the whole issue.
Consider this case:
  you previously exported a `.csv`. You click export on the same data again, switch the filter to `PNG Image (*.png)` and hit `Save`. What do you expect will happen?~

I was wrong, this is implemented exactly as described in `QFileDialog`, I take it all back, fix incoming

File: pywisp/gui.py (hunk 3 above)
Comment (jhalmen) [reply pywisp/gui.py:1207]: huh, don't remember what I must have been doing wrong, last week, but moving the waiting for a thread into a method and calling it normally actually works...

fix incoming

Comment (cklb) [reply pywisp/gui.py:1208]: I like it. Having Exporter.wait() saves outsiders from directly interacting with the internal worker Thread.

**Comment by cklb** [comment]
Very nice. I updated the exporter tests to make use of the new methods and added some comments about the behavior being tested.

I also (very professionally) clicked around some more in the plot and export widgets and was not able to find a new bug.

I think we are good to go :smile:

**Comment by jhalmen** [comment]
I agree
